File: csharp/send_notification_with_action.cs
using System;
using System.Net.Http;
using System.Text;
// https://www.nuget.org/packages/Newtonsoft.Json/
using Newtonsoft.Json.Linq;
// https://www.nuget.org/packages/Nito.AsyncEx/
using Nito.AsyncEx;

namespace csharp
{
    class Program
    {
        // Android doc -> http://docs.pushe.co/docs/mobile-api/send_notification/
        // Obtain token -> http://docs.pushe.co/docs/mobile-api/authentication/
        static String token = "YOUR_TOKEN";

        static void Main(string[] args)
        {
            AsyncContext.Run(() => sendNotification());
        }

        public static async void sendNotification()
        {
            var client = new HttpClient();
            client.DefaultRequestHeaders.Add("Authorization", "Token " + token);

            var response = await client.PostAsync(
                "https://api.pushe.co/v2/messaging/notifications/",
                getNotificationData()
            );
            // In order to send a notification to iOS applications use this url
            // https://api.pushe.co/v2/messaging/notifications/ios

            var status_code = (int)response.StatusCode;
            var reponse_json = JObject.Parse(await response.Content.ReadAsStringAsync());

            Console.WriteLine("status code => " + status_code);
            Console.WriteLine("response => " + reponse_json.ToString());
            Console.WriteLine("==========");

            if (status_code == 201)
            {
                Console.WriteLine("Success!");

                var hashed_id = reponse_json.GetValue("hashed_id").ToString();
                String report_url;

                if (String.IsNullOrEmpty(hashed_id))
                {
                    report_url = "no report url for your plan";
                }
                else
                {
                    report_url = "https://pushe.co/report?id=" + hashed_id;
                }
                Console.WriteLine("report_url: " + report_url);

                var notif_id = reponse_json.GetValue("wrapper_id").ToString();
                Console.WriteLine("notification id: " + notif_id);
            }
            else
            {
                Console.WriteLine("failed!");
            }
        }

        public static StringContent getNotificationData()
        {
            var data = new JObject();
            data.Add("title", "This is a notification with buttons");
            data.Add("content", "In this notification, every button has an action.");

            var action = new JObject();
            action.Add("action_type", "U");
            action.Add("url", "https://pushe.co");

            var openAppAction = new JObject();
            openAppAction.Add("action_type", "A");

            var tellAction = new JObject();
            tellAction.Add("action_type", "U");
            tellAction.Add("url", "tel:02187654321");

            var marketActionParams = new JObject();
            marketActionParams.Add("market", "bazaar");
            marketActionParams.Add("package_name", "shop.barkat.app");
            
            var installAppAction = new JObject();
            installAppAction.Add("action_type", "U");
            installAppAction.Add("url", "bazaar://details?id=shop.barkat.app");
            installAppAction.Add("market_package_name", "com.farsitel.bazaar");
            installAppAction.Add("params", marketActionParams);

            var openAppBtn = new JObject();
            openAppBtn.Add("btn_content", "Open App");
            openAppBtn.Add("btn_order", 0);
            openAppBtn.Add("btn_action", openAppAction);

            var tellBtn = new JObject();
            tellBtn.Add("btn_content", "YOUR_CONTENT");
            tellBtn.Add("btn_order", 1);
            tellBtn.Add("btn_action", tellAction);

            var installAppBtn = new JObject();
            installAppBtn.Add("btn_content", "Install App");
            installAppBtn.Add("btn_order", 1);
            installAppBtn.Add("btn_action", installAppAction);
            
            data.Add("action",action);
            data.add("buttons",new JArray(new JObject[]{openAppBtn,tellBtn,installAppBtn}));
            
            var request_data = new JObject();
            request_data.Add("app_ids", new JArray(new String[] { "YOUR_APPLICATION_ID" }));
            request_data.Add("data", data);


            Console.WriteLine("Request data: " + request_data.ToString());

            return new StringContent(request_data.ToString(), Encoding.UTF8, "application/json");
        }
    }
}

File: php/send_notification_with_action.php
<?php

// Android doc -> http://docs.pushe.co/docs/mobile-api/notification-actions/
// Obtain token -> http://docs.pushe.co/docs/mobile-api/authentication/
$TOKEN = "YOUR_TOKEN";

$data = array(
    "app_ids" => ["YOUR_APP_ID"],
    "data" => array(
        "title" => "This is a notification with buttons",
        "content" => "In this notification, every button has an action.",
        "action" => array(
            "url" => "",
            "action_type" => "A"
        ),
        "buttons"=> array(
            array(
                "btn_content"=> "Open App",
                "btn_action"=> array("action_type"=> "A"),
                "btn_order"=> 0),
            ),
            array(
                "btn_content"=> "Call",
                "btn_action"=> array(
                    "action_type"=> "U",
                    "url"=> "tel:02187654321"
                ),
                "btn_order"=> 1
            ),
            array(
                "btn_content"=> "Install App",
                "btn_action"=> array(
                    "action_type"=> "U",
                    "params"=> array("market"=> "bazaar", "package_name"=> "shop.barkat.app"),
                    "url"=> "bazaar://details?id=shop.barkat.app",
                    "market_package_name"=> "com.farsitel.bazaar"
                ),
                "btn_icon"=> "add_box",
                "btn_order"=> 2
            )
        )
    );

$ch = curl_init("https://api.pushe.co/v2/messaging/notifications/");
// In order to send a notification to iOS applications use this url
// https://api.pushe.co/v2/messaging/notifications/ios

curl_setopt($ch, CURLOPT_HTTPHEADER, array(
    "Content-Type: application/json",
    "Accept: application/json",
    "Authorization: Token " . $TOKEN,
));

curl_setopt($ch, CURLOPT_POST, true);
curl_setopt($ch, CURLOPT_POSTFIELDS, json_encode($data));
curl_setopt($ch, CURLOPT_RETURNTRANSFER, true);

$reponse = curl_exec($ch);
$status_code = curl_getinfo($ch, CURLINFO_HTTP_CODE);
curl_close($ch);

echo "status code => $status_code\n";
echo "response => $reponse\n";
echo "==========\n";

if ($status_code == 201) {
    echo "Success!\n";

    $reponse_json = json_decode($reponse);

    if ($reponse_json->hashed_id) {
        $report_url = "https://pushe.co/report?id=$reponse_json->hashed_id";
    } else {
        $report_url = "no report url for your plan";
    }
    echo "report_url: $report_url\n";

    echo "notification id: $reponse_json->wrapper_id\n";
} else {
    echo "failed!\n";
}
